File: scripts/workflow_brief.py
#!/usr/bin/env python3
"""Build a compact execution brief from the IPD-centered workflow state."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any

import workflow_guard


def extract_task_headings(text: str) -> list[str]:
    headings: list[str] = []
    for line in text.splitlines():
        stripped = line.strip()
        if stripped.startswith("## ") and "TASK-" in stripped:
            headings.append(stripped.removeprefix("## ").strip())
    return headings


def build_brief(root: Path | str, *, feature: str, doc_id: str, target: str = "execute") -> dict[str, Any]:
    root_path = Path(root)
    guard = workflow_guard.check(root_path, feature=feature, doc_id=doc_id, target=target)
    next_context = guard["next_context"]
    must_read = next_context["must_read"]

    task_headings: list[str] = []
    if must_read:
        ipd_path = root_path / must_read[0]
        if ipd_path.exists():
            task_headings = extract_task_headings(ipd_path.read_text(encoding="utf-8"))

    return {
        "pass": guard["pass"],
        "feature": feature,
        "doc_id": doc_id,
        "target": target,
        "state": guard["state"],
        "reason": guard["reason"],
        "next_skill": guard["next_skill"],
        "failures": guard["failures"],
        "must_read": must_read,
        "may_skip": next_context["may_skip"],
        "focus_sections": next_context["focus_sections"],
        "task_headings": task_headings,
        "execution_source": next_context["execution_source"],
        "new_plan_policy": "create-new-ipd",
    }


def format_plain(payload: dict[str, Any]) -> str:
    lines = [
        f"State: {payload['state']}",
        f"Pass: {str(payload['pass']).lower()}",
        f"Reason: {payload['reason']}",
        f"Next skill: {payload['next_skill']}",
        "Must read:",
    ]
    lines.extend(f"- {path}" for path in payload["must_read"] or ["-"])
    lines.append("May skip unless rewind triggers fire:")
    lines.extend(f"- {path}" for path in payload["may_skip"] or ["-"])
    if payload["task_headings"]:
        lines.append("Task chapters:")
        lines.extend(f"- {heading}" for heading in payload["task_headings"])
    if payload["failures"]:
        lines.append("Failures:")
        lines.extend(f"- {failure}" for failure in payload["failures"])
    lines.append(f"Execution source: {payload['execution_source']}")
    lines.append(f"New plan policy: {payload['new_plan_policy']}")
    return "\n".join(lines)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Build a compact Coding Plugins workflow brief.")
    parser.add_argument("--root", default=".")
    parser.add_argument("--feature", required=True)
    parser.add_argument("--doc-id", required=True)
    parser.add_argument("--target", choices=sorted(workflow_guard.VALID_TARGETS), default="execute")
    parser.add_argument("--json", action="store_true")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    payload = build_brief(Path(args.root), feature=args.feature, doc_id=args.doc_id, target=args.target)
    if args.json:
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    else:
        print(format_plain(payload))
    return 0 if payload["pass"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: src/lib/workflow-brief.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

import { checkWorkflowGuard, VALID_TARGETS } from "./workflow-guard.ts";

export { VALID_TARGETS };

interface BriefOptions {
  feature: string;
  docId: string;
  target?: string;
  task?: string;
}

export interface WorkflowBriefPayload {
  pass: boolean;
  feature: string;
  doc_id: string;
  target: string;
  current_task?: string | null;
  state: string;
  reason: string;
  next_skill: string;
  failures: string[];
  must_read: string[];
  may_skip: string[];
  focus_sections: string[];
  task_headings: string[];
  execution_source: string;
  new_plan_policy: string;
}

export function extractTaskHeadings(text: string): string[] {
  const headings: string[] = [];
  for (const line of text.split(/\r?\n/)) {
    const stripped = line.trim();
    if (stripped.startsWith("## ") && stripped.includes("TASK-")) {
      headings.push(stripped.slice("## ".length).trim());
    }
  }
  return headings;
}

export function buildBrief(root: string, options: BriefOptions): WorkflowBriefPayload {
  const target = options.target ?? "execute";
  if (!VALID_TARGETS.has(target)) {
    throw new Error(`invalid workflow guard target: ${target}`);
  }

  const guard = checkWorkflowGuard(root, {
    feature: options.feature,
    docId: options.docId,
    target,
  });
  const nextContext = guard.next_context;
  const mustRead = nextContext.must_read;

  let taskHeadings: string[] = [];
  if (mustRead.length > 0) {
    const planPath = join(root, mustRead[0]);
    if (existsSync(planPath)) {
      taskHeadings = extractTaskHeadings(readFileSync(planPath, "utf8"));
    }
  }

  const failures = [...guard.failures];
  const focusSections = [...nextContext.focus_sections];
  if (options.task) {
    taskHeadings = taskHeadings.filter((heading) => heading.includes(options.task ?? ""));
    if (taskHeadings.length > 0) {
      focusSections.push(...taskHeadings.map((heading) => `## ${heading}`));
    } else {
      failures.push(`requested task ${options.task} was not found`);
    }
  }

  return {
    pass: guard.pass && failures.length === 0,
    feature: options.feature,
    doc_id: options.docId,
    target,
    current_task: options.task ?? null,
    state: guard.state,
    reason: guard.reason,
    next_skill: guard.next_skill,
    failures,
    must_read: mustRead,
    may_skip: nextContext.may_skip,
    focus_sections: focusSections,
    task_headings: taskHeadings,
    execution_source: nextContext.execution_source,
    new_plan_policy: "create-new-ted",
  };
}

export function formatPlain(payload: WorkflowBriefPayload): string {
  const lines = [
    `State: ${payload.state}`,
    `Pass: ${String(payload.pass).toLowerCase()}`,
    `Reason: ${payload.reason}`,
    `Next skill: ${payload.next_skill}`,
    "Must read:",
  ];
  lines.push(...(payload.must_read.length > 0 ? payload.must_read : ["-"]).map((path) => `- ${path}`));
  lines.push("May skip unless rewind triggers fire:");
  lines.push(...(payload.may_skip.length > 0 ? payload.may_skip : ["-"]).map((path) => `- ${path}`));
  if (payload.task_headings.length > 0) {
    lines.push("Task chapters:");
    lines.push(...payload.task_headings.map((heading) => `- ${heading}`));
  }
  if (payload.failures.length > 0) {
    lines.push("Failures:");
    lines.push(...payload.failures.map((failure) => `- ${failure}`));
  }
  lines.push(`Execution source: ${payload.execution_source}`);
  lines.push(`New plan policy: ${payload.new_plan_policy}`);
  return lines.join("\n");
}
